source/common/http/filter/cors_filter.cc

```diff
@@ -0,0 +1,152 @@
+#include "common/http/filter/cors_filter.h"
+
+#include "envoy/http/codes.h"
+
+#include "common/common/enum_to_int.h"
+#include "common/http/header_map_impl.h"
+#include "common/http/headers.h"
+
+namespace Envoy {
+namespace Http {
+
+CorsFilter::CorsFilter(CorsFilterConfigConstSharedPtr config)
+    : config_(config), is_cors_request_(false) {}
+
+CorsFilter::~CorsFilter() {}
+
+FilterHeadersStatus CorsFilter::decodeHeaders(HeaderMap& headers, bool) {
+  origin_ = headers.Origin();
+  if (origin_ == nullptr || origin_->value() == "") {
+    return FilterHeadersStatus::Continue;
+  }
+
+  initialize();
+
+  if (!enabled()) {
+    return FilterHeadersStatus::Continue;
+  }
+
+  if (!isOriginAllowed(origin_->value())) {
+    return FilterHeadersStatus::Continue;
+  }
+
+  is_cors_request_ = true;
+
+  auto method = headers.Method();
+  if (method == nullptr || method->value().c_str() != Http::Headers::get().MethodValues.Options) {
+    return FilterHeadersStatus::Continue;
+  }
+
+  auto requestMethod = headers.AccessControlRequestMethod();
+  if (requestMethod == nullptr || requestMethod->value() == "") {
+    return FilterHeadersStatus::Continue;
+  }
+
+  HeaderMapPtr response_headers{
+      new HeaderMapImpl{{Headers::get().Status, std::to_string(enumToInt(Http::Code::OK))}}};
+
+  response_headers->insertAccessControlAllowOrigin().value(*origin_);
+
+  if (allowCredentials()) {
+    response_headers->insertAccessControlAllowCredentials().value(
+        Http::Headers::get().CORSValues.True);
+  }
+
+  if (!allowMethods().empty()) {
+    response_headers->insertAccessControlAllowMethods().value(allowMethods());
+  }
+
+  if (!allowHeaders().empty()) {
+    response_headers->insertAccessControlAllowHeaders().value(allowHeaders());
+  }
+
+  if (!exposeHeaders().empty()) {
+    response_headers->insertAccessControlExposeHeaders().value(exposeHeaders());
+  }
+
+  if (!maxAge().empty()) {
+    response_headers->insertAccessControlMaxAge().value(maxAge());
+  }
+
+  decoder_callbacks_->encodeHeaders(std::move(response_headers), true);
+
+  return FilterHeadersStatus::StopIteration;
+}
+
+FilterHeadersStatus CorsFilter::encodeHeaders(HeaderMap& headers, bool) {
```

Comment:
I don't think this filter needs to be an encoder filter. Is there any reason to make it one? You already respond in decodeHeaders() above so why not just do everything there? Then we don't to store extra state inside the filter itself in member variables.

Reply:
I found that after the preflight request, which is handled in decode, every response needs the `allow-origin` header or the browser won't accept it. That's why I add it to the upstream response.

Reply:
I mean every non-preflight cors request needs that header set.

Reply:
Ah OK. Makes sense. I will review again post-update with that in mind.

Reply:
Can you add a small comment here on why we need to do this in encodeHeaders()? (Either link to spec or just tiny explanation for future readers).